sql/core/src/main/scala/org/apache/spark/sql/ui/SQLListener.scala

```diff
@@ -0,0 +1,340 @@
+                // heartbeat reports
+                taskMetrics.accumulatorUpdates = metrics.accumulatorUpdates()
+              }
+            case None =>
+              // TODO Now just set attemptId to 0. Should fix here when we can get the attempt
+              // id from SparkListenerExecutorMetricsUpdate
+              stageMetrics.taskIdToMetricUpdates(taskId) =
+                SQLTaskMetrics(attemptId = 0, finished = finishTask, metrics.accumulatorUpdates())
+          }
+        }
+      case None =>
+      // This execution and its stage have been dropped
+    }
+  }
+
+  def onExecutionStart(
+      executionId: Long,
+      description: String,
+      details: String,
+      df: DataFrame,
+      time: Long): Unit = {
+    val physicalPlanDescription = df.queryExecution.toString
+    val physicalPlanGraph = SparkPlanGraph(df.queryExecution.executedPlan)
+    val metrics = physicalPlanGraph.nodes.flatMap { node =>
+      node.metrics.map(metric => metric.accumulatorId -> metric)
+    }
+
+    val executionUIData = SQLExecutionUIData(executionId, description, details,
+      physicalPlanDescription, physicalPlanGraph, metrics.toMap, time)
+
+    synchronized {
+      activeExecutions(executionId) = executionUIData
+      executionIdToData(executionId) = executionUIData
+    }
+  }
+
+  def onExecutionEnd(executionId: Long, time: Long): Unit = synchronized {
+    executionIdToData.get(executionId).foreach { executionUIData =>
+      executionUIData.completionTime = Some(time)
+      if (!executionUIData.hasRunningJobs) {
+        // onExecutionEnd happens after all "onJobEnd"s
+        // So we should update the execution lists.
+        updateExecutionLists(executionId)
+      } else {
+        // There are some running jobs, onExecutionEnd happens before some "onJobEnd"s.
+        // Then we don't if the execution is successful, so let the last onJobEnd updates the
+        // execution lists.
+      }
+    }
+  }
+
+  private def updateExecutionLists(executionId: Long): Unit = {
+    activeExecutions.remove(executionId).foreach { executionUIData =>
+      if (executionUIData.isFailed) {
+        failedExecutions += executionUIData
+        trimExecutionsIfNecessary(failedExecutions)
+      } else {
+        completedExecutions += executionUIData
+        trimExecutionsIfNecessary(completedExecutions)
+      }
+    }
+  }
+
+  def getRunningExecutions: Seq[SQLExecutionUIData] = synchronized {
+    activeExecutions.values.toSeq
+  }
+
+  def getFailedExecutions: Seq[SQLExecutionUIData] = synchronized {
+    failedExecutions
+  }
+
+  def getCompletedExecutions: Seq[SQLExecutionUIData] = synchronized {
+    completedExecutions
+  }
+
+  def getExecution(executionId: Long): Option[SQLExecutionUIData] = synchronized {
+    executionIdToData.get(executionId)
+  }
+
+  def getExecutionMetrics(executionId: Long): Map[Long, Any] = synchronized {
```

Comment:
can you add a java doc to this one? It's really returning the values of the metrics, right?